holoviews/streams.py

```diff
@@ -0,0 +1,166 @@
+"""
+The streams module defines the streams API that allows visualizations to
+generate and respond to events, originating either in Python on the
+server-side or in Javascript in the Jupyter notebook (client-side).
+"""
+
+import param
+import uuid
+from collections import OrderedDict
+
+
+class Stream(param.Parameterized):
+    """
+    A Stream is simply a parameterized object with parameters that
+    change over time in response to update events. Parameters are
+    updated via the update method.
+
+    Streams may have one or more subscribers, callables that are passed
+    the parameter dictionary when the trigger classmethod is called.
+    """
+
+    # Mapping from uuid to stream instance
+    registry = OrderedDict()
+
+    @classmethod
+    def trigger(cls, streams):
+        """
+        Given a list of streams, collect all the stream parameters into
+        a dictionary and pass it to the union set of subscribers.
+
+        Passing multiple streams at once to trigger can be useful when a
+        subscriber may be set multiple times across streams but only
+        needs to be called once.
+        """
+        # Union of stream values
+        items = [stream.value.items() for stream in streams]
+        union = dict(kv for kvs in items for kv in kvs)
+        # Currently building a simple set of subscribers
+        groups = [stream.subscribers + stream._hidden_subscribers for stream in streams]
+        subscribers = set(s for subscribers in groups for s in subscribers)
+        for subscriber in subscribers:
+            subscriber(union)
+
+    def update(self, trigger=True, **kwargs):
+        """
+        The update method updates the stream parameters in response to
+        some event.
+
+        If trigger is enabled, the trigger classmethod is invoked on
+        this particular Stream instance.
+        """
+        for param in self.params().values():
+            param.constant = False
+        self.set_param(**kwargs)
+        for param in self.params().values():
+            param.constant = True
```

Comment:
So parameters can't be set directly, ever? How would this play with paramnb, where the widgets set the parameter values directly. I was hoping I could simply attach Stream.update as a callback on the paramnb widgets and have it just work, but if the params are constant it will never work.